master/pkg/ws/ws.go

```diff
@@ -0,0 +1,262 @@
+package ws
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/hashicorp/go-multierror"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// pingInterval is the interval at which to send pings.
+	pingInterval = 15 * time.Second
+	// pongWait is the duration to wait for a pong response to a ping.
+	pongWait = time.Minute
+	// closeWait is the duration to wait for a close response.
+	closeWait = 5 * time.Second
+	// inboxBufferSize is the number of messages to read before applying backpressure.
+	inboxBufferSize = 32
+	// outboxBufferSize is the number of messages to write before applying backpressure.
+	outboxBufferSize = 64
+	// maxMessageSize is the maximum size of a websocket message that we send in bytes, copied
+	// from MAX_WEBSOCKET_MSG_SIZE in determined/constants.py.
+	maxMessageSize = 128 * 1024 * 1024
+)
+
+// Websocket provides a higher level websocket library on top of "github.com/gorilla/websocket".
+type Websocket[TIn, TOut any] struct {
+	// System dependencies.
+	log  *logrus.Entry
+	conn *websocket.Conn
+
+	// Internal state.
+	cancel    context.CancelFunc
+	errLock   sync.Mutex
+	err       error
+	closeOnce sync.Once
+	closeErr  error
+	// Done signals the websocket is finished when it is closed. Even if it has exited, call Close.
+	Done <-chan struct{}
+	// Inbox is a channel for incoming messages.
+	Inbox <-chan TIn
+	// Outbox is a channel for outgoing messages.
+	Outbox chan<- TOut
+}
+
+// Wrap the given, underlying *websocket.Conn and returns a higher level, thread-safe wrapper.
+func Wrap[TIn, TOut any](name string, conn *websocket.Conn) *Websocket[TIn, TOut] {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	inbox := make(chan TIn, inboxBufferSize)
+	outbox := make(chan TOut, outboxBufferSize)
+	done := make(chan struct{})
+
+	s := &Websocket[TIn, TOut]{
+		log: logrus.WithFields(logrus.Fields{
+			"component":   "websocket",
+			"remote-addr": conn.RemoteAddr(),
+			"name":        name,
+		}),
+		conn: conn,
+
+		cancel: cancel,
+		Done:   done,
+		Inbox:  inbox,
+		Outbox: outbox,
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := s.runWriteLoop(ctx, outbox); err != nil {
+			s.setError(fmt.Errorf("write loop: %w", err))
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := s.runReadLoop(ctx, inbox); err != nil {
+			s.setError(fmt.Errorf("read loop: %w", err))
+		}
+	}()
+
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	return s
+}
+
+func (s *Websocket[TIn, TOut]) Wait() error {
```

Comment:
Why have `s.Done` exported and also define this method? If folks want a one-liner for blocking and then reading the error, they can do something like
```go
_, wsErr := <-ws.Done, ws.Error()
```
[example](https://go.dev/play/p/cXIJ4ML9MR4)

Reply:
Yeah, true. Though I don't think there is any harm in providing .wait. But I don't think it is used anyway, outside tests, so I'll drop it.